### PR TITLE
Recommendation 3.9 split enum

### LINKED ISSUE #153: Recommendation 3.9: enum values are separated by comma and a line break. ';' is put on the new line
1) Enum values are separated by comma and a line break. ';' is put on the new line:
     ```Kotlin
        enum class Warnings {
            A,
            B,
            C
            ;
        }
     ```
    This will help to resolve conflicts and reduce it's number during merging pull requests.

2) In case enum is simple (it has no properties, no methods and no comments inside) - it can be declared in a single line:
     ```kotlin
    enum class Suit { CLUBS, HEARTS, SPADES, DIAMONDS }
     ```

3) Prefer enum classes if it is possible, for example instead of 2 Boolean properties like:
    `val isCelsius = true | val isFahrenheit = false` use enum class:
    ```kotlin
    enum class TemperatureScale { CELSIUS, FAHRENHEIT }
    ```

    - The variable value only changes within a fixed range and is defined with the enum type. For example, [Recommend 8.6.3 Keyboard](#s8.6.3) example
    - Avoid comparison with magic constant numbers of -1, 0, and 1, instead of this use enums:
  ```kotlin
    enum class ComparisonResult {
        ORDERED_ASCENDING,
        ORDERED_SAME,
        ORDERED_DESCENDING
        ;
    }
  ```

### LINKED ISSUE #196: Recommendation 3.9: enum values are separated by comma and a line break. ';' is put on the new line
Recommendation 3.9: enum values are separated by comma and a line break. ';' is put on the new line
1) Enum values are separated by comma and after last value must be comma and  a line break. ';' is put on the new line:

```kotlin
   enum class Warnings {
       A,
       B,
       C,
       ;
   }
```
This will help to resolve conflicts and reduce it's number during merging pull requests.

2) In case enum is simple (it has no properties, no methods and no comments inside) - it can be declared in a single line:

```kotlin
enum class Suit { CLUBS, HEARTS, SPADES, DIAMONDS }
```
3) Prefer enum classes if it is possible, for example instead of 2 Boolean properties like: val isCelsius = true | val isFahrenheit = false use enum class:

```kotlin
enum class TemperatureScale { CELSIUS, FAHRENHEIT }
```
The variable value only changes within a fixed range and is defined with the enum type. For example, Recommend 8.6.3 Keyboard example
Avoid comparison with magic constant numbers of -1, 0, and 1, instead of this use enums:

```kotlin
  enum class ComparisonResult {
      ORDERED_ASCENDING,
      ORDERED_SAME,
      ORDERED_DESCENDING
      ;
  }
```

### PR DESCRIPTION
**Pull request by kentr0w**
## Which rule and warnings did you add?
Recommendation 3.9: enum values are separated by comma and a line break. ';' is put on the new line

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Added tests on fixers
* [x] Updated rules-config.json
* [x] Updated available-rules.md
* [x] closes #196 and closes #153 

## Fixme
    Prefer enum classes if it is possible, for example instead of 2 Boolean properties like: val isCelsius = true | val isFahrenheit = false use enum class:

### CODE REVIEW

File: diktat-rules/src/test/resources/test/paragraph3/enum_separated/EnumSeparatedExpected.kt [hunk 1]
```diff
@@ -41,3 +41,22 @@ enum class ProtocolState {
...
 }
+
+enum class Simple {
+    A, B, C
+}
+
+enum class SimpleWithFun {
+    A,
+ B,
+ C ,
```
Comment (petertrr) [inline diktat-rules/src/test/resources/test/paragraph3/enum_separated/EnumSeparatedExpected.kt:52]: Could you add the comma node right after identifier, without this space?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                ENUMS_SEPARATED.warnAndFix(configRules, emitWarn, isFixMode, "enum constance must end with a line break",
+                        it.startOffset) {
+                    it.appendNewlineMergingWhiteSpace(it.treeNext, it.treeNext)
+                }
+        }
+        checkLastEnum(enums.last())
+    }
+
+    private fun isSimpleEnumOneLine(nodes: List<ASTNode>) =
+            nodes.find { it != nodes.last() && it.treeNext.isWhiteSpaceWithNewline() } == null
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:65]:
```suggestion
            nodes.dropLast(1).find { it.treeNext.isWhiteSpaceWithNewline() } == null
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        isFixMode = autoCorrect
+
+        if (node.elementType == CLASS)
+            if (node.isClassEnum())
+                checkEnumEntry(node)
+    }
+
+    //Fixme prefer enum classes if it is possible instead of variables
+    private fun checkEnumEntry(node: ASTNode) {
+        val enums = node.findChildByType(CLASS_BODY)!!.getAllChildrenWithType(ENUM_ENTRY)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:51]: Please rename this variable for readability to something like `enumEntries`.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                    it.appendNewlineMergingWhiteSpace(it.treeNext, it.treeNext)
+                }
+        }
+        checkLastEnum(enums.last())
+    }
+
+    private fun isSimpleEnumOneLine(nodes: List<ASTNode>) =
+            nodes.find { it != nodes.last() && it.treeNext.isWhiteSpaceWithNewline() } == null
+
+    private fun isEnumSimple(nodes: List<ASTNode>): Boolean {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:67]: Please rename `nodes` argument for readability to something like `enumEntries`.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt [hunk 5]
```diff
@@ -230,6 +230,11 @@ fun ASTNode.findAllNodesWithSpecificType(elementType: IElementType): List<ASTNod
     }
 }
 
+/**
+ * Check a node of type CLASS if it is a enum class
+ */
+fun ASTNode.isClassEnum(): Boolean = this.findChildByType(MODIFIER_LIST)?.hasChildOfType(ElementType.ENUM_KEYWORD) ?: false
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt:236]:
```suggestion
fun ASTNode.isClassEnum(): Boolean = (psi as? KtClass)?.isEnum() ?: false
```
What do you think?

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/EnumsSeparatedWarnTest.kt [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+class EnumsSeparatedWarnTest {
+
+    private val ruleId = "$DIKTAT_RULE_SET_ID:enum-separated"
+
+    @Test
+    @Tag(WarningNames.ENUMS_SEPARATED)
+    fun `check simple correct enum with new line`() {
+        lintMethod(EnumsSeparated(),
+                """
+                    |enum class ENUM {
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/EnumsSeparatedWarnTest.kt:21]: Please also add example with comments on each line, for example this should be valid:
```kotlin
enum class ENUM {
   A, // this is A
   B,
   ;
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                node.addChild(PsiWhiteSpaceImpl("\n"), node.findChildByType(SEMICOLON)!!)
+            }
+        } else if (!node.findChildByType(SEMICOLON)!!.treePrev.isWhiteSpaceWithNewline()) {
+            ENUMS_SEPARATED.warnAndFix(configRules, emitWarn, isFixMode, "semicolon must be on a new line",
+                    node.startOffset) {
+                node.appendNewlineMergingWhiteSpace(node.findChildByType(SEMICOLON)!!, node.findChildByType(SEMICOLON)!!)
+            }
+        }
+        if (!node.hasChildOfType(COMMA)) {
+            ENUMS_SEPARATED.warnAndFix(configRules, emitWarn, isFixMode, "last enum constance must end with a comma",
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:89]: What do you mean by 'constance'? I think it's more common to call them 'enum entries'.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            if (!it.treeNext.isWhiteSpaceWithNewline())
+                ENUMS_SEPARATED.warnAndFix(configRules, emitWarn, isFixMode, "enum constance must end with a line break",
+                        it.startOffset) {
+                    it.appendNewlineMergingWhiteSpace(it.treeNext, it.treeNext)
+                }
+        }
+        checkLastEnum(enums.last())
+    }
+
+    private fun isSimpleEnumOneLine(nodes: List<ASTNode>) =
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:64]:
```suggestion
    private fun isEnumOneLine(nodes: List<ASTNode>) =
```
This check is independent from `isSimple` and doesn't contain this logic.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt (hunk 5 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt:236]: I think it much better. I didn't know about this.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       params: KtLint.Params,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        configRules = params.getDiktatConfigRules()
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == CLASS)
+            if (node.isClassEnum())
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:45]: I would suggest to also check if the node has child of type `CLASS_BODY`. *Theoretically*, empty enum class is valid in kotlin, and though it is strange and not really useful, it's probably worth checking.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                    it.appendNewlineMergingWhiteSpace(it.treeNext, it.treeNext)
+                }
+        }
+        checkLastEnum(enumEntries.last())
+    }
+
+    private fun isSimpleEnumOneLine(nodes: List<ASTNode>) =
+            nodes.dropLast(1).find { it.treeNext.isWhiteSpaceWithNewline() } == null
+
+    private fun isEnumOneLine(enumEntries: List<ASTNode>): Boolean {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:67]: But this method doesn't check if enum is an one line, it checks if it is simple.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        enumEntries.forEach { node ->
+            if (!SIMPLE_VALUE.containsAll(node.getChildren(null).map { it.elementType }))
+                return false
+        }
+        return SIMPLE_ENUM.containsAll(enumEntries.last().treeParent.getChildren(null).map { it.elementType })
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:72]:
```suggestion
        return SIMPLE_ENUM.containsAll(enumEntries.last().allSiblings(withSelf = true).map { it.elementType })
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                node.addChild(PsiWhiteSpaceImpl("\n"), node.findChildByType(SEMICOLON)!!)
+            }
+        } else if (!node.findChildByType(SEMICOLON)!!.treePrev.isWhiteSpaceWithNewline()) {
+            ENUMS_SEPARATED.warnAndFix(configRules, emitWarn, isFixMode, "semicolon must be on a new line",
+                    node.startOffset) {
+                node.appendNewlineMergingWhiteSpace(node.findChildByType(SEMICOLON)!!, node.findChildByType(SEMICOLON)!!)
+            }
+        }
+        if (!node.hasChildOfType(COMMA)) {
+            ENUMS_SEPARATED.warnAndFix(configRules, emitWarn, isFixMode, "last enum entries must end with a comma",
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:89]:
```suggestion
            ENUMS_SEPARATED.warnAndFix(configRules, emitWarn, isFixMode, "last enum entry must end with a comma",
```

File: info/available-rules.md [hunk 13]
```diff
@@ -62,4 +62,5 @@
...
-|    3 |      3.8 | TOO_MANY_CONSECUTIVE_SPACES             | Check: warns if there are too many consecutive spaces in line. Exception: in enum if configured and single eol comments <br>Fix: squeezes spaces to 1.                       | yes | max_spaces saveInitialFormattingForEnums | - |
\ No newline at end of file
+|    3 |      3.8 | TOO_MANY_CONSECUTIVE_SPACES             | Check: warns if there are too many consecutive spaces in line. Exception: in enum if configured and single eol comments <br>Fix: squeezes spaces to 1.                       | yes | max_spaces saveInitialFormattingForEnums | - |
+|    3 |      3.9 | ENUMS_SEPARATED                          | Check: warns if enum structure is correct: separated by comma and line break and semicolon in the and.                                                                       | yes |                                                                                                                                                                                                            | -                         | Replace variable to enum if it possible
```
Comment (petertrr) [inline info/available-rules.md:66]:
```suggestion
|    3 |      3.9 | ENUMS_SEPARATED                          | Check: warns if enum structure is incorrect: enum entries should be separated by comma and line break and last entry should have semicolon in the end.                                                                       | yes |                                                                                                                                                                                                            | -                         | Replace variable to enum if it possible |
```

**Review by petertrr — CHANGES_REQUESTED** [review]

**Review by petertrr — CHANGES_REQUESTED** [review]
I think there has been an error during merge in `available-rules.md`. 'Tests' column has been removed on master, and your PR re-adds it. Please check.

File: info/available-rules.md [hunk 14]
```diff
@@ -62,4 +62,5 @@
...
-|    3 |      3.8 | TOO_MANY_CONSECUTIVE_SPACES             | Check: warns if there are too many consecutive spaces in line. Exception: in enum if configured and single eol comments <br>Fix: squeezes spaces to 1.                       | yes | max_spaces saveInitialFormattingForEnums | - |
\ No newline at end of file
+|    3 |      3.8 | TOO_MANY_CONSECUTIVE_SPACES             | Check: warns if there are too many consecutive spaces in line. Exception: in enum if configured and single eol comments <br>Fix: squeezes spaces to 1.                       | yes | max_spaces saveInitialFormattingForEnums | - |
+|    3 |      3.9 | ENUMS_SEPARATED                         | Check: warns if enum structure is incorrect: enum entries should be separated by comma and line break and last entry should have semicolon in the end.                       | yes |-                          | Replace variable to enum if it possible
```
Comment (petertrr) [inline info/available-rules.md:66]:
```suggestion
|    3 |      3.9 | ENUMS_SEPARATED                         | Check: warns if enum structure is incorrect: enum entries should be separated by comma and line break and last entry should have semicolon in the end.                       | yes |-                          | Replace variable to enum if it possible |
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+                ENUMS_SEPARATED.warnAndFix(configRules, emitWarn, isFixMode, "enum entries must end with a line break",
+                        it.startOffset) {
+                    it.appendNewlineMergingWhiteSpace(it.treeNext, it.treeNext)
+                }
+        }
+        checkLastEnum(enumEntries.last())
+    }
+
+    private fun isEnumOneLine(nodes: List<ASTNode>) =
+            nodes.dropLast(1).find { it.treeNext.isWhiteSpaceWithNewline() } == null
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/EnumsSeparated.kt:66]:
```suggestion
            nodes.dropLast(1).none { it.treeNext.isWhiteSpaceWithNewline() }
```

**Review by petertrr — APPROVED** [review]
LGTM, just two more really minor comments.

**Pull request merged**